recipes/z3/all/conanfile.py

```diff
@@ -71,20 +82,25 @@ def _configure_cmake(self):
         return self._cmake
 
     def build(self):
-        for patch in self.conan_data.get("patches", {}).get(self.version, []):
+        for patch in self.conan_data.get("patches").get(self.version):
+            if patch["patch_file"] == "0001-cmake-use-conan-mpir.patch" and self.options.multiprecision != "mpir":
```

Comment:
This is not stable code because it depends on exact strings and the path is actually `"patches/0001-cmake-use-conan-mpir-4.8.8.patch"`.
Instead I propose to modify the patch to do something as follows:
```cmake
if (Z3_USE_LIB_GMP)
# Because this is off by default we will make the configure fail if libgmp
# can't be found
essage(STATUS "Using libgmp")
if(USE_GMP)
  find_package(GMP REQUIRED)
  list(APPEND Z3_DEPENDENT_LIBS GMP::GMP)
else()
  find_package(mpir REQUIRED)
  list(APPEND Z3_DEPENDENT_LIBS mpir::mpir)
endif()
list(APPEND Z3_COMPONENT_CXX_DEFINES "-D_MP_GMP")
else()
```

Reply:
Still not stable enough IMO.
The best would be not to touch the CMakeLists.txt at all.

Why did you need to do that here, instead of patching `z3`?
Is it because the `z3` does not want to do it that way?

Reply:
And why `mpir` instead of their internal code?

Reply:
mpir is a fork of gmp, and is fully compatible so no issue there.

When a project links to both a static z3 and  a static mpir/gmp, the linker might mix z3's private multiprecision implementation and mpir/gmp, which causes undefined behavior.

Reply:
The patch would need to be modified for the different versions of z3, if required.
Most often, it is only a matter of line numbers that have changed

Reply:
Yes, so how is your alternative piece of code better in terms of stability?

Reply:
I've just understood: you're talking about the string in the Python script that should remain the same. But it can't, it will depend on the versions. I have to deal with the `conandata.yml` more cleverly.

Reply:
1. You're using cmake functionality (variables + if/else)
2. Changing the name of the patch file won't change the behavior of the build unexpectedly

Reply:
Please don't conditionally appy patches or depend on the patch filename.
It's really prone to breakage.

Reply:
Thanks, I get it now.
I'm going along your proposition now.
But how about patching the `z3` repository so that it is easy to build with `mpir`?